Modules/_cryptmodule.c

```diff
@@ -1,66 +1,93 @@
-will be used to perturb the encryption algorithm and produce distinct
-results for a given *word*.
-
-[clinic start generated code]*/
+PyDoc_STRVAR(crypt_crypt__doc__,
+"crypt($module, word, salt, /)\n"
+"--\n"
+"\n"
+"Hash a *word* with the given *salt* and return the hashed password.\n"
+"\n"
+"*word* will usually be a user\'s password.  *salt* (either a random 2 or 16\n"
+"character string, possibly prefixed with $digit$ to indicate the method)\n"
+"will be used to perturb the encryption algorithm and produce distinct\n"
+"results for a given *word*.");
 
 static PyObject *
-crypt_crypt_impl(PyObject *module, const char *word, const char *salt)
-/*[clinic end generated code: output=0512284a03d2803c input=0e8edec9c364352b]*/
+crypt_crypt(PyObject *module, PyObject *const *args, Py_ssize_t nargs)
 {
+    if (nargs != 2) {
+        PyErr_Format(PyExc_TypeError,"crypt expected 2 arguments, got %zd", nargs);
+        return NULL;
+    }
+    const char *word;
+    const char *salt;
+    Py_ssize_t word_length;
+    Py_ssize_t salt_length;
+
+    if (!PyUnicode_Check(args[0])) {
+        PyErr_Format(
+            PyExc_TypeError, "crypt argument 1 (word) must be string, not %s",
+            Py_TYPE(args[0])->tp_name
+        );
+        return NULL;
+    }
+    word = PyUnicode_AsUTF8AndSize(args[0], &word_length);
+    if (word == NULL) {
+        return NULL;
+    }
+    if (strlen(word) != (size_t)word_length) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "crypt argument 1 (word) contains embedded null character"
+        );
+        return NULL;
+    }
+
+    if (!PyUnicode_Check(args[1])) {
+        PyErr_Format(
+            PyExc_TypeError, "crypt argument 2 (salt) must be string, not %s",
+            Py_TYPE(args[1])->tp_name
+        );
+        return NULL;
+    }
+    salt = PyUnicode_AsUTF8AndSize(args[1], &salt_length);
+    if (salt == NULL) {
+        return NULL;
+    }
+    if (strlen(salt) != (size_t)salt_length) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "crypt argument 2 (salt) contains embedded null character"
+        );
+        return NULL;
+    }
+
     char *crypt_result;
-#ifdef HAVE_CRYPT_R
     struct crypt_data data;
     memset(&data, 0, sizeof(data));
     crypt_result = crypt_r(word, salt, &data);
-#else
-    crypt_result = crypt(word, salt);
-#endif
     if (crypt_result == NULL) {
         return PyErr_SetFromErrno(PyExc_OSError);
     }
-    return Py_BuildValue("s", crypt_result);
+    return PyUnicode_FromString(crypt_result);
 }
 
 
 static PyMethodDef crypt_methods[] = {
-    CRYPT_CRYPT_METHODDEF
+    {"crypt", crypt_crypt, METH_FASTCALL, crypt_crypt__doc__},
     {NULL,              NULL}           /* sentinel */
 };
 
 static PyModuleDef_Slot _crypt_slots[] = {
-#ifdef Py_MOD_PER_INTERPRETER_GIL_SUPPORTED
-    {Py_mod_multiple_interpreters, Py_MOD_PER_INTERPRETER_GIL_SUPPORTED},
-#endif
```

Comment:
What is the thought behind this removal?

Reply:
IIRC, @encukou mentioned it used to be there for older pythons.

Reply:
It was added quite recently: https://github.com/python/cpython/commit/a9c6e0618f26270e2591b3d99ffeef55eea02a33

Reply:
It is also documented and does not seem internal:

https://docs.python.org/3/c-api/module.html#c.Py_MOD_PER_INTERPRETER_GIL_SUPPORTED

Let's keep it for now and we can remove it later if it proves problematic... ?

Reply:
Ah, the removal would make sense for `crypt` (not `crypt_r`), which is not thread-safe.

But this PR now removes the call to `crypt` (because we can't rely on `HAVE_CRYPT_R` which comes from Python's `configure`), so it should be safe to bring `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` back.